firmware/application/apps/ble_app.cpp

```diff
@@ -0,0 +1,430 @@
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, Inc., 51 Franklin Street,
+ * Boston, MA 02110-1301, USA.
+ */
+
+#include "ble_app.hpp"
+#include "ui_modemsetup.hpp"
+
+#include "modems.hpp"
+#include "audio.hpp"
+#include "rtc_time.hpp"
+#include "baseband_api.hpp"
+#include "string_format.hpp"
+#include "portapack_persistent_memory.hpp"
+
+#define USE_CONSOLE 0
+
+using namespace portapack;
+using namespace modems;
+
+void BLELogger::log_raw_data(const std::string& data) {
+    log_file.write_entry(data);
+}
+
+std::string padStringWithSpaces(int n) {
+    std::string paddedStr(n, ' ');
+    return paddedStr;
+}
+
+namespace ui {
+template <>
+void RecentEntriesTable<BleRecentEntries>::draw(
+    const Entry& entry,
+    const Rect& target_rect,
+    Painter& painter,
+    const Style& style) {
+    std::string line = to_string_hex(entry.macAddress & 0xFF, 2);
+
+    line += ":" + to_string_hex((entry.macAddress >> 8) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 16) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 24) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 32) & 0xFF, 2);
+    line += ":" + to_string_hex((entry.macAddress >> 40), 2);
+
+    // Handle spacing for negitive sign.
+    uint8_t db_spacing = entry.dbValue > 0 ? 7 : 6;
+
+    // Pushing single digit values down right justified.
+    if (entry.dbValue > 9 || entry.dbValue < -9) {
+        db_spacing--;
+    }
+
+    line += padStringWithSpaces(db_spacing) + to_string_dec_int(entry.dbValue);
+
+    line.resize(target_rect.width() / 8, ' ');
+    painter.draw_string(target_rect.location(), style, line);
+}
+
+BleRecentEntryDetailView::BleRecentEntryDetailView() {
+    add_children({&button_done,
+                  &labels});
+
+    button_done.on_select = [this](const ui::Button&) {
+        if (on_close) {
+            on_close();
+        }
+    };
+}
+
+BleRecentEntryDetailView::BleRecentEntryDetailView(const BleRecentEntryDetailView& Entry)
+    : View() {
+    (void)Entry;
+}
+
+BleRecentEntryDetailView& BleRecentEntryDetailView::operator=(const BleRecentEntryDetailView& Entry) {
+    (void)Entry;
+    return *this;
+}
+
+void BleRecentEntryDetailView::update_data() {
+}
+
+void BleRecentEntryDetailView::focus() {
+    button_done.focus();
+}
+
+Rect BleRecentEntryDetailView::draw_field(
+    Painter& painter,
+    const Rect& draw_rect,
+    const Style& style,
+    const std::string& label,
+    const std::string& value) {
+    const int label_length_max = 4;
+
+    painter.draw_string(Point{draw_rect.left(), draw_rect.top()}, style, label);
+    painter.draw_string(Point{draw_rect.left() + (label_length_max + 1) * 8, draw_rect.top()}, style, value);
+
+    return {draw_rect.left(), draw_rect.top() + draw_rect.height(), draw_rect.width(), draw_rect.height()};
+}
+
+void BleRecentEntryDetailView::paint(Painter& painter) {
+    View::paint(painter);
```

Comment:
I haven't see what this looks like to know whether you really need custom drawing here. Typically, views don't paint themselves. Usually widgets paint themselves and views hold widgets and everything just works. If you actually have something custom to draw, that usually indicates that you want to also add a widget to encapsulate the drawing of a single "item".
Can you post a screenshot of this?

Reply:
I think I see what you mean. I copied this from ais, but maybe ais is wrong too? I could somehow make it a Widget, but I haven't made one yet so I was unsure.

Reply:
Could be, this project has gone through quite a few hands at this point and not all of the older stuff has been brought forward.
It's not a blocker for commit, I just wanted to point out the how the UI architecture works so you can build your mental model of what belongs where as you work on the code.